### PR TITLE
rados: make all rados op w/ FULL_TRY flag.

### PR DESCRIPTION
**Pull request by majianpeng**
Like rados command 'rm' --force-full.

### CODE REVIEW

File: src/librados/IoCtxImpl.cc [hunk 1]
```diff
@@ -1005,6 +1005,29 @@ int librados::IoCtxImpl::aio_remove(const object_t &oid, AioCompletionImpl *c)
   return 0;
 }
 
+int librados::IoCtxImpl::aio_remove(const object_t &oid, AioCompletionImpl *c, int flags)
```
Comment (yangdongsheng) [inline src/librados/IoCtxImpl.cc]: Then we can change the original aio_remove to
```
int librados::IoCtxImpl::aio_remove(const object_t &oid, AioCompletionImpl *c)
{
    return aio_remove(oid, c, 0);
}
```

File: src/tools/rados/rados.cc [hunk 2]
```diff
@@ -69,8 +69,8 @@ void usage(ostream& out)
 "   cppool <pool-name> <dest-pool>   copy content of a pool\n"
 "   rmpool <pool-name> [<pool-name> --yes-i-really-really-mean-it]\n"
 "                                    remove pool <pool-name>'\n"
-"   purge <pool-name> --yes-i-really-really-mean-it\n"
-"                                    remove all objects from pool <pool-name> without removing it\n"
+"   purge <pool-name> --yes-i-really-really-mean-it [--force-full]\n"
+"                                    [force no mater full or not]remove all objects from pool <pool-name> without removing it\n"
```
Comment (dzafman) [inline src/tools/rados/rados.cc]: s/mater/matter/

File: src/tools/rados/rados.cc [hunk 3]
```diff
@@ -101,8 +101,8 @@ void usage(ostream& out)
 "                                    default is 16 concurrent IOs and 4 MB ops\n"
 "                                    default is to clean up after write benchmark\n"
 "                                    default run-name is 'benchmark_last_metadata'\n"
-"   cleanup [--run-name run_name] [--prefix prefix]\n"
-"                                    clean up a previous benchmark operation\n"
+"   cleanup [--run-name run_name] [--prefix prefix] [--force-full]\n"
+"                                    [force no mater full or not]clean up a previous benchmark operation\n"
```
Comment (dzafman) [inline src/tools/rados/rados.cc]: s/mater/matter/

File: src/tools/rados/rados.cc [hunk 4]
```diff
@@ -903,7 +904,10 @@ class RadosBencher : public ObjBencher {
   }
 
   int aio_remove(const std::string& oid, int slot) {
-    return io_ctx.aio_remove(oid, completions[slot]);
+    if (forcefull)
+      return io_ctx.aio_remove(oid, completions[slot], CEPH_OSD_FLAG_FULL_FORCE);
+    else
+      return io_ctx.aio_remove(oid, completions[slot]);
```
Comment (dzafman) [inline src/tools/rados/rados.cc]: sync_remove() needs this equivalent change too calling the flags version of IoCtx::remove().  This removes the "benchmark_last_metadata" object.

Comment (dzafman) [inline src/tools/rados/rados.cc]: Maybe forceflag state should set an int flags and then just have one call to aio_remove:

<pre>
return io_ctx::aoi_remove(oid, completes[slot], flags);
</pre>

**Comment by majianpeng** [comment]
@dzafman . update.

**Comment by dzafman** [comment]
Either squash the 2 commits, or change the description to reflect the rados command commit.

**Comment by dzafman** [comment]
I tried your branch and it didn't work because of an existing bug in the handling of CEPH_OSD_FLAG_FULL_FORCE flag.

Please cherry-pick the following commit into your branch: c5a4c6c4149e7d4fd38aa2a6f20c591af71cc5c3

**Comment by majianpeng** [comment]
@dzafman,  I changed ceph.conf:  osd failsafe full ratio = .30; mon osd full ratio = .10, The first two patch work. But if mon osd full ratio = osd failsafe full ratio. It can't work. So i add e2414ff.

In MOSDOp *Objecter::_prepare_osd_op(Op *op) 
>> if (!honor_osdmap_full)
 >>   flags |= CEPH_OSD_FLAG_FULL_FORCE;

But only  mds call  unset_honor_osdmap_full(). So for FULL_FORCE, we suppose use know the result of this.It should do whether osd has full flag or osd is failsafe.

**Comment by majianpeng** [comment]
@liewegas . This mainly for how to handle FULL_TRY/FULL_FORCE when osd in full or failsafe statue. Hope your comments. Thanks!

File: src/librados/IoCtxImpl.h [hunk 5]
```diff
@@ -206,6 +206,7 @@ struct librados::IoCtxImpl {
   int aio_writesame(const object_t &oid, AioCompletionImpl *c,
 		    const bufferlist& bl, size_t write_len, uint64_t off);
   int aio_remove(const object_t &oid, AioCompletionImpl *c);
+  int aio_remove(const object_t &oid, AioCompletionImpl *c, int flags);
```
Comment (liewegas) [inline src/librados/IoCtxImpl.h]: @jdurgin adding a new overload breaks the ABI, right?  this should be aio_remove2?

Should we just use aio_operate from the cli instead?

File: src/osd/ReplicatedPG.cc [hunk 6]
```diff
@@ -6631,8 +6630,7 @@ int ReplicatedPG::prepare_transaction(OpContext *ctx)
       (pool.info.has_flag(pg_pool_t::FLAG_FULL) ||
        get_osdmap()->test_flag(CEPH_OSDMAP_FULL))) {
     MOSDOp *m = static_cast<MOSDOp*>(ctx->op->get_req());
-    if (ctx->reqid.name.is_mds() ||   // FIXME: ignore MDS for now
-	m->has_flag(CEPH_OSD_FLAG_FULL_FORCE)) {
+    if (m->has_flag(CEPH_OSD_FLAG_FULL_FORCE)) {
       dout(20) << __func__ << " full, but proceeding due to FULL_FORCE or MDS"
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: remove ' or MDS'

**Comment by liewegas** [comment]
The problem here is that the MDS can/will now bypass the failsafe and break an OSD (by overfilling it such that it can't start).  I wonder if the right thing to do is to make the MDS set FULL_TRY instead of FULL_FORCE?

@gregsfortytwo @jcsp thoughts?  It's probably time we had a real plan for ENOSPC here.

**Comment by gregsfortytwo** [comment]
@liewegas unless I'm misremembering, the MDS at this point will only submit delete operations when the osdmap full flag is set; everything else gets an ENOSPC. I guess that still leaves a race between when the OSD actually fills up and when the MDS gets a map saying stuff is full, but that's what the thresholds are supposed to be for. :/

Or else I'm misremembering how much of that is done as I believe @jcsp did most of it.

File: src/librados/IoCtxImpl.h (hunk 5 above)
Comment (jdurgin) [inline src/librados/IoCtxImpl.h]: yeah, I'd rather use aio_operate() from the cli instead of adding a new method to librados.

**Comment by liewegas** [comment]
@majianpeng So, for the rados CLI issue, I *think* what you actaully want is FULL_TRY.  That will make the OSD prepare the transaction and only allow it if it doesn't increase utilization.. which is enough for the CLI to allow deletes (utilization will clearly go down).

File: src/tools/rados/rados.cc [hunk 7]
```diff
@@ -69,8 +69,8 @@ void usage(ostream& out)
 "   cppool <pool-name> <dest-pool>   copy content of a pool\n"
 "   rmpool <pool-name> [<pool-name> --yes-i-really-really-mean-it]\n"
 "                                    remove pool <pool-name>'\n"
-"   purge <pool-name> --yes-i-really-really-mean-it\n"
-"                                    remove all objects from pool <pool-name> without removing it\n"
+"   purge <pool-name> --yes-i-really-really-mean-it [--force-full]\n"
+"                                    [force no matter full or not]remove all objects from pool <pool-name> without removing it\n"
```
Comment (liewegas) [inline src/tools/rados/rados.cc]: please drop the [...] part

File: src/tools/rados/rados.cc [hunk 8]
```diff
@@ -101,8 +101,8 @@ void usage(ostream& out)
 "                                    default is 16 concurrent IOs and 4 MB ops\n"
 "                                    default is to clean up after write benchmark\n"
 "                                    default run-name is 'benchmark_last_metadata'\n"
-"   cleanup [--run-name run_name] [--prefix prefix]\n"
-"                                    clean up a previous benchmark operation\n"
+"   cleanup [--run-name run_name] [--prefix prefix] [--force-full]\n"
+"                                    [force no matter full or not]clean up a previous benchmark operation\n"
```
Comment (liewegas) [inline src/tools/rados/rados.cc]: please drop the [..] part

File: src/tools/rados/rados.cc [hunk 9]
```diff
@@ -903,7 +904,8 @@ class RadosBencher : public ObjBencher {
   }
 
   int aio_remove(const std::string& oid, int slot) {
-    return io_ctx.aio_remove(oid, completions[slot]);
+    int flags = forcefull ? CEPH_OSD_FLAG_FULL_FORCE : 0;
```
Comment (liewegas) [inline src/tools/rados/rados.cc]: I think this should be FULL_TRY, not FULL_FORCE.

**Comment by dzafman** [comment]
@jcsp @gregsfortytwo @liewegas Let's verify that MDS uses the FULL_TRY flag in all the correct place before turning off the is_mds() checking?

File: src/osd/ReplicatedPG.cc [hunk 10]
```diff
@@ -1673,17 +1673,17 @@ void ReplicatedPG::do_op(OpRequestRef& op)
...
-  // FIXME: we exclude mds writes for now.
-  if (!(m->get_source().is_mds() || m->has_flag(CEPH_OSD_FLAG_FULL_FORCE)) &&
+  if (!m->has_flag(CEPH_OSD_FLAG_FULL_FORCE|CEPH_OSD_FLAG_FULL_TRY) &&
       info.history.last_epoch_marked_full > m->get_map_epoch()) {
     dout(10) << __func__ << " discarding op sent before full " << m << " "
 	     << *m << dendl;
     return;
   }
-  if (!m->get_source().is_mds() && osd->check_failsafe_full()) {
+  if (!m->has_flag(CEPH_OSD_FLAG_FULL_FORCE|CEPH_OSD_FLAG_FULL_FORCE) &&
```
Comment (dzafman) [inline src/osd/ReplicatedPG.cc]: This needs to be CEPH_OSD_FLAG_FULL_FORCE|CEPH_OSD_FLAG_FULL_TRY.  I wish this code looked like this:

<pre>
if (!m->has_flag(CEPH_OSD_FLAG_FULL_FORCE|CEPH_OSD_FLAG_FULL_TRY) {
    if (info.history.last_epoch_marked_full > m->get_map_epoch()) {
        dout(...
        return;
    }
    if (osd->check_failsafe_full()) {
        dout(...
        return;
    }
}
</pre>

**Comment by liewegas** [comment]
@majianpeng ok, having thought at this some more, I think:

1. first patch: don't add a new aio_remove op.  Users can use the aio_operate instead and set the flag that way.
2. for the rados cli, I think we can just uncondutionally set FULL_TRY on the rados op.  It just means the op will get an error instead of blocking, which I think is what we always want from the cli.
3. The failsafe check should remove the mds special case--it's dangerous!
4. the last two patches look correct to me as is.

Thank you!

**Comment by gregsfortytwo** [comment]
@dzafman, we don't use FULL_TRY at all in the MDS. I was just making a statement about when it is willing to submit operations to the Objecter. :) Just created a feature ticket, though: http://tracker.ceph.com/issues/15942

**Comment by dzafman** [comment]
@liewegas 

> 2. for the rados cli, I think we can just uncondutionally set FULL_TRY on the rados op. It just means the op will get an error instead of blocking, which I think is what we always want from the cli.

Yes, maybe the cli should set FULL_TRY on all operations.  The only reason not to do that is for scripting that wants to block.  For object removal --force-full sets FULL_FORCE so that space can be freed up.

Proposal:
Default rados cli operations sets FULL_TRY (e.g. "put" and "import" can fail with ENOSPC)
NEW --block-full does not set FULL_TRY so that scripts can hang on out of space
--force-full Object remove operations would pass FULL_FORCE if this option specified

**Comment by majianpeng** [comment]
To make all rados op w/ FULL_TRY, the simplest method is add new api like objecter::set_honor_osdmap_full. I think for all client this simple. So  I add Objecter::set_osdmap_full_op_try(). To be frankly, the name of api is ugly. Hope suggestion.

In fact, at first, i wonder change honor_osdmap_full flag from FULL_FORCE to FULL_TRY. I think FULL_TRY more general and FULL_FORCE maybe set for  individual op. Hope suggestions.

Thanks!

**Comment by liewegas** [comment]
Hmm, yeah.  Let's

1. Do a cleanup patch that renames honor_osdmap_full to osdmap_full_force. This will flip the bool around and change the Objecter interface to set_osdmap_full_force(bool f)
2. Add a the set_osdmap_full_try(bool t).
3. Call the librados API set_block_on_full(bool block) where it defaults to true, and false means we set FULL_TRY.  Ths has to be a Rados method, though, not IoCtx, or else it implies it's only affects that IoCtx.  (And that would require changing all of the IoCtx.cc operate() etc calls to pass the flag explicitly instead of setting it on Objecter.  That might be nice, but it's a lot more work and i don't think we need it. Yet.)

**Comment by liewegas** [comment]
Also, I think this was unclear before, but we should drop 'osd/ReplicatedPG: Handle op which w/ FULL_FORCE if osd is failsafe. …'.  It should no longer be possible to bypass the failsafe check at all.

**Comment by majianpeng** [comment]
@liewegas . update

File: src/osd/ReplicatedPG.cc [hunk 11]
```diff
@@ -1654,17 +1654,18 @@ void ReplicatedPG::do_op(OpRequestRef& op)
...
-  if (!(m->get_source().is_mds() || m->has_flag(CEPH_OSD_FLAG_FULL_FORCE)) &&
+  if (!(m->get_source().is_mds() || m->has_flag(CEPH_OSD_FLAG_FULL_FORCE|CEPH_OSD_FLAG_FULL_TRY)) &&
       info.history.last_epoch_marked_full > m->get_map_epoch()) {
     dout(10) << __func__ << " discarding op sent before full " << m << " "
 	     << *m << dendl;
     return;
   }
-  if (!m->get_source().is_mds() && osd->check_failsafe_full()) {
+  if (!(m->get_source().is_mds() || m->has_flag(CEPH_OSD_FLAG_FULL_TRY)) &&
+      osd->check_failsafe_full()) {
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: this line shouldn't change--we shouldn't skip the failsafe check.

File: src/osd/ReplicatedPG.cc [hunk 12]
```diff
@@ -1657,14 +1657,13 @@ void ReplicatedPG::do_op(OpRequestRef& op)
...
-  if (!(m->get_source().is_mds() || m->has_flag(CEPH_OSD_FLAG_FULL_FORCE|CEPH_OSD_FLAG_FULL_TRY)) &&
+  if (!(m->has_flag(CEPH_OSD_FLAG_FULL_FORCE|CEPH_OSD_FLAG_FULL_TRY)) &&
       info.history.last_epoch_marked_full > m->get_map_epoch()) {
     dout(10) << __func__ << " discarding op sent before full " << m << " "
 	     << *m << dendl;
     return;
   }
-  if (!(m->get_source().is_mds() || m->has_flag(CEPH_OSD_FLAG_FULL_TRY)) &&
+  if (!(m->has_flag(CEPH_OSD_FLAG_FULL_TRY)) &&
       osd->check_failsafe_full()) {
```
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: so that here we should end up with just if (osd->check_failsafe_full()) { ... }

File: src/osdc/Objecter.cc [hunk 13]
```diff
@@ -2310,22 +2310,32 @@ void Objecter::_op_submit(Op *op, shunique_lock& sul, ceph_tid_t *ptid)
...
+  /*
+   * For FULL_FORCE/FULL_TRY, it can set directly on OP or set by Objcter
+   * like osdmap_full_force & osdmap_full_try
+   */
+  int flags = op->target.flags;
+  if (osdmap_full_force)
+    flags |= CEPH_OSD_FLAG_FULL_FORCE;
+  if (osdmap_full_try)
+    flags |= CEPH_OSD_FLAG_FULL_TRY;
+
```
Comment (liewegas) [inline src/osdc/Objecter.cc]: i think this should just set the bit in op->target.flags, and _prepare_osd_op() can drop the part the sets it when sending.

**Comment by majianpeng** [comment]
@liewegas . Does it need add set_block_on_osdfull for librados c-api?

**Comment by liewegas** [comment]
we probably should!

**Comment by majianpeng** [comment]
@liewegas  add api for librados c-api: rados_set_block_on_osdfull.

File: src/osd/ReplicatedPG.cc (hunk 12 above)
Comment (liewegas) [inline src/osd/ReplicatedPG.cc]: this line still needs to be fixed.  the client should not be able to bypass the failsafe check.

**Comment by liewegas** [comment]
looks good, except for that failsafe check condition (the client should not be able to bypass it with the full_try flag).

**Comment by jcsp** [comment]
This appears to be causing an fs test failure:
http://qa-proxy.ceph.com/teuthology/yuriw-2016-06-07_18:47:13-fs-wip-yuri-testing_JUN_7_2016---basic-smithi/243224/teuthology.log

**Comment by majianpeng** [comment]
@jcsp . I think this because 	a3b4bdd. If osd->check_failsafe_full() return true, for the delete op, it will discard and don't anything. Although the op w/ FULL_FORCE flag.
@liewegas,  i think for FULL_FORCE, it wall bypass check_failsafe_full.

**Comment by athanatos** [comment]
This seemed to cause a crash in sjust@teuthology:/a/yuriw-2016-06-14_09:06:01-rados-wip-yuri-testing_JUN_7_2016-distro-basic-smithi/259449

**Comment by majianpeng** [comment]
For my test, change osd_failsafe_full_ratio = 0.10; mon_osd_full_ratio=0.99; if osd reach the failsafe_full; all ops will discard include rm op. This because sage's suggestion.  At before 2f30a97. mds op can bypass this state.  From the log, i think it don't remove the object and so don't add new object(because osd now in failsafe full state)

**Comment by athanatos** [comment]
@majianpeng I don't understand.  If you need to change the values for the test, do it, and submit a PR for ceph-qa-suite as well.  We can't merge this if it breaks existing tests.

**Comment by jcsp** [comment]
@majianpeng are you still working on this?

**Comment by dachary** [comment]
needs rebase

**Comment by dachary** [comment]
feel free to re-open once you have time to work on it

**Pull request closed without merging**